tests/unit/completion/test_sqlmodel.py

```diff
@@ -0,0 +1,240 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2016 Ryan Roden-Corrent (rcorre) <ryan@rcorre.net>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Tests for the base sql completion model."""
+
+import pytest
+from PyQt5.QtCore import Qt
+
+from qutebrowser.misc import sql
+from qutebrowser.completion.models import sqlmodel
+
+
+pytestmark = pytest.mark.usefixtures('init_sql')
+
+
+def _check_model(model, expected):
+    """Check that a model contains the expected items in the given order.
+
+    Args:
+        expected: A list of form
+                  [
+                      (cat, [(name, desc, misc), (name, desc, misc), ...]),
+                      (cat, [(name, desc, misc), (name, desc, misc), ...]),
+                      ...
+                  ]
+    """
+    assert model.rowCount() == len(expected)
+    for i, (expected_title, expected_items) in enumerate(expected):
+        catidx = model.index(i, 0)
+        assert model.data(catidx) == expected_title
+        assert model.rowCount(catidx) == len(expected_items)
+        for j, (name, desc, misc) in enumerate(expected_items):
+            assert model.data(model.index(j, 0, catidx)) == name
+            assert model.data(model.index(j, 1, catidx)) == desc
+            assert model.data(model.index(j, 2, catidx)) == misc
+
+
+@pytest.mark.parametrize('rowcounts, expected', [
+    ([0], 0),
+    ([1], 1),
+    ([2], 2),
+    ([0, 0], 0),
+    ([0, 0, 0], 0),
+    ([1, 1], 2),
+    ([3, 2, 1], 6),
+    ([0, 2, 0], 2),
+])
+def test_count(rowcounts, expected):
+    model = sqlmodel.SqlCompletionModel()
+    for i, rowcount in enumerate(rowcounts):
+        name = 'Foo' + str(i)
+        table = sql.SqlTable(name, ['a'], primary_key='a')
+        for rownum in range(rowcount):
+            table.insert([rownum])
+        model.new_category(name)
```

Comment:
Described above, but basically everthing is referencing the same in-memory sql database. If something (e.g. the `BookmarkManager` creates a table named "Bookmarks", then something else (e.g. a completion model) can access that table given nothing more than it's name.